File: app/src/main/java/com/arny/myapidemo/mvp/useredit/UserEditFragment.java
package com.arny.myapidemo.mvp.useredit;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ImageView;
import com.afollestad.materialdialogs.MaterialDialog;
import com.arellomobile.mvp.MvpAppCompatFragment;
import com.arellomobile.mvp.presenter.InjectPresenter;
import com.arny.myapidemo.R;
import com.bumptech.glide.Glide;

public class UserEditFragment extends MvpAppCompatFragment implements UserEditView {
    @InjectPresenter
    UserEditPresenter userEditPresenter;
    private ImageView imgAvatar;
    private EditText edtLogin;
    private EditText editEmail;
    private EditText edtPass;
    private EditText edtPassConfirm;
    private ImageButton imgBtnEditAvatar;
    private ImageButton imgBtnDeleteAvatar;
    private FloatingActionButton fabEdit;
    private Context context;
    private MaterialDialog materialDialog;

    public UserEditFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        this.context = container.getContext();
        return inflater.inflate(R.layout.fragment_user_edit, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        imgAvatar = view.findViewById(R.id.imgAvatar);
        edtLogin = view.findViewById(R.id.edtLogin);
        editEmail = view.findViewById(R.id.edtEmail);
        edtPass = view.findViewById(R.id.edtPass);
        edtPassConfirm = view.findViewById(R.id.edtPassConfirm);
        imgBtnEditAvatar = view.findViewById(R.id.imgBtnEditAvatar);
        imgBtnDeleteAvatar = view.findViewById(R.id.imgBtnDeleteAvatar);
        fabEdit = view.findViewById(R.id.fabEdit);
        if (getArguments() != null) {
            String login = getArguments().getString("login");
            userEditPresenter.initState(context, login);
        }
    }

    @Override
    public void initUI(String login, String email, String avatar) {
        edtLogin.setText(login);
        editEmail.setText(email);
        if (avatar != null) {
            Glide.with(context).load(avatar).into(imgAvatar);
        }
    }

    @Override
    public void hideErrors() {

    }

    @Override
    public void setToolbarTitle(String title) {

    }

    @Override
    public void updateProgress(String progress) {

    }

    @Override
    public void showAlert(String message) {
        if (materialDialog == null || !materialDialog.isShowing()) {
            materialDialog = new MaterialDialog.Builder(context)
                    .title(R.string.app_name)
                    .content(message)
                    .positiveText(android.R.string.ok)
                    .onPositive((dialog, which) -> userEditPresenter.alertRead())
                    .cancelable(false)
                    .show();
        }
    }

    @Override
    public void hideMessage() {
        if (materialDialog != null && materialDialog.isShowing()) {
            materialDialog.dismiss();
        }
    }
}

File: app/src/main/java/com/arny/myapidemo/mvp/useredit/UserEditFragment.kt
package com.arny.myapidemo.mvp.useredit

import android.os.Bundle
import android.os.Environment
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.afollestad.materialdialogs.MaterialDialog
import com.arellomobile.mvp.MvpAppCompatFragment
import com.arellomobile.mvp.presenter.InjectPresenter
import com.arellomobile.mvp.presenter.PresenterType
import com.arny.myapidemo.R
import com.bumptech.glide.Glide
import kotlinx.android.synthetic.main.fragment_user_edit.*
import com.afollestad.materialdialogs.folderselector.FileChooserDialog


class UserEditFragment : MvpAppCompatFragment(), UserEditView, View.OnClickListener {
    override fun viewImage(uri: String?) {
        Log.i("UserEditFragment","uri:" + uri)
        val loaded = Glide.with(context).asBitmap().load(uri)
        Log.i("UserEditFragment","loaded:" + loaded)
        Glide.with(context)
                .load(uri)
                .into(imgAvatar)
    }

    @InjectPresenter
    var userEditPresenter: UserEditPresenter? = null
    private var materialDialog: MaterialDialog? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_user_edit, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        fabEdit.setOnClickListener(this)
        imgBtnEditAvatar.setOnClickListener(this)
        if (arguments != null) {
            val login = arguments!!.getString("login")
            context?.let { userEditPresenter?.initState(it, login) }
        }
    }

    override fun initUI(login: String, email: String, avatar: String?) {
        edtLogin.setText(login)
        edtEmail.setText(email)
        if (avatar != null) {
            Glide.with(context).load(avatar).into(imgAvatar!!)
        }
    }

    override fun hideErrors() {

    }

    override fun setToolbarTitle(title: String) {
        if (activity != null) {
            activity!!.title = title
        }
    }

    override fun updateProgress(progress: String) {

    }

    override fun showAlert(message: String) {
        if (materialDialog == null || !materialDialog!!.isShowing) {
            materialDialog = MaterialDialog.Builder(context!!)
                    .title(R.string.app_name)
                    .content(message)
                    .positiveText(android.R.string.ok)
                    .onPositive { dialog, which -> userEditPresenter!!.alertRead() }
                    .cancelable(false)
                    .show()
        }
    }

    override fun hideMessage() {
        if (materialDialog != null && materialDialog!!.isShowing) {
            materialDialog!!.dismiss()
        }
    }

    override fun onClick(view: View) {
        when (view.id) {
            R.id.btnSignIn -> context?.let { userEditPresenter?.register(edtLogin.text.toString(), edtPass.text.toString(), it) }
            R.id.imgBtnEditAvatar -> {
                context?.let {
                    FileChooserDialog.Builder(it)
                            .initialPath(Environment.getExternalStorageDirectory().getPath())
                            .mimeType("image/*")
                            .extensionsFilter(".png", ".jpg") // Optional extension filter, will override mimeType()
                            .tag("select_image")
                            .goUpLabel("Up") // custom go up label, default label is "..."
                            .show(activity)
                }
            }
        }
    }
}
